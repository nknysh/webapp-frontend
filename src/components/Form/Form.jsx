import React, { forwardRef } from 'react';
import { Formik } from 'formik';

import { propTypes } from './Form.props';

export const Form = forwardRef(({ children, ...props }, ref) => (
  <Formik ref={ref} {...props}>
    {({ handleSubmit, ...formProps }) => <form onSubmit={handleSubmit}>{children(formProps)}</form>}
  </Formik>
));

Form.displayName = 'Form';
Form.propTypes = propTypes;

export default Form;
