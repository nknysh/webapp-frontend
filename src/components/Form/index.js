// Libraries
import React from 'react';
import { Formik } from 'formik';

const Form = ({ children, ...props }) => <Formik {...props} render={children} />;

export default Form;
