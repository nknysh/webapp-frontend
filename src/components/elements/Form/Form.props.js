import PropTypes from 'prop-types';

export const propTypes = {
  children: PropTypes.func,
  fields: PropTypes.array,
};

export const defaultProps = {
  fields: [],
};
