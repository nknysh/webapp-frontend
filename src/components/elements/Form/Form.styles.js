import styled from 'styled-components';

import ServerError from 'components/elements/ServerError';
import { theme } from 'styles';

export const ServerErrorContent = styled(ServerError)`
  margin-top: ${theme.gutter * 2}px;
`;
