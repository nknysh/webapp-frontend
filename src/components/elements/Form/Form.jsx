import React, { forwardRef } from 'react';
import { Formik } from 'formik';
import { mapObjIndexed, values, partialRight, path } from 'ramda';

import { getFormPath } from 'utils';

import FormField from 'components/elements/FormField';
import FormFieldError from 'components/elements/FormFieldError';

import { propTypes, defaultProps } from './Form.props';
import { ServerErrorContent } from './Form.styles';

const renderFormError = content => <ServerErrorContent>{content}</ServerErrorContent>;

const renderError = error => error && <FormFieldError>{error}</FormFieldError>;

const renderField = (name, value, field, { handleChange, handleBlur, errors }, checked) => (
  <FormField
    name={name}
    value={value}
    checked={checked}
    onChange={handleChange}
    onBlur={handleBlur}
    error={path(getFormPath(name), errors)}
    {...field}
  />
);

const renderFields = (field, name, formProps) => renderField(name, undefined, field, formProps);

export const Form = forwardRef(({ children, fields, ...props }, ref) => (
  <Formik ref={ref} {...props}>
    {({ handleSubmit, ...formProps }) => (
      <form onSubmit={handleSubmit}>
        {children(formProps)}
        {values(mapObjIndexed(partialRight(renderFields, [formProps]), fields))}
      </form>
    )}
  </Formik>
));

Form.displayName = 'Form';
Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

Form.renderField = renderField;
Form.renderFieldError = renderError;
Form.renderFormError = renderFormError;

export default Form;
