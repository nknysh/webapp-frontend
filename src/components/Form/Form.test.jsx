import React from 'react';

import { Form } from './Form';

const getComponent = props => shallow(<Form {...props} />);

describe('<Form />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
