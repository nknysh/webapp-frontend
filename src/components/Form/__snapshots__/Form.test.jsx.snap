// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> render matches snapshot 1`] = `
<Formik
  enableReinitialize={false}
  isInitialValid={false}
  validateOnBlur={true}
  validateOnChange={true}
>
  <Component />
</Formik>
`;
